### PR TITLE
Jinja Variant Configuration

### PR DESCRIPTION
**Pull request by alanmcruickshank**
This makes progress on #4061 . I got a bit burnt out before so I'm trying to keep the steps small.

This introduces the basic configuration for linting variants, and crosses the boundary into the linter. That also enables the `sqlfluff render` cli command to show rendered variants which is now supported.

image

It defaults to using `process_with_variants` for all templaters, and so should trigger a minor release because it will be a breaking change for anyone using custom templaters. Quite a few type hints were inaccurate - which is one of the reasons this has been a little convoluted. I've straightened out a bunch of them in the process of this work.

Unless someone _explicitly_ sets a configuration value to enable this - nothing is enabled. The default value for `sqlfluff:render_variant_limit` is 1 which effectively disables it unless changed.

### CODE REVIEW

File: src/sqlfluff/core/linter/linter.py [hunk 1]
```diff
@@ -714,23 +723,34 @@ def render_string(
...
+
         try:
-            templated_file, templater_violations = self.templater.process(
+            for variant, vs in self.templater.process_with_variants(
                 in_str=in_str, fname=fname, config=config, formatter=self.formatter
-            )
+            ):
+                if variant:
+                    templated_variants.append(variant)
+                templater_violations += vs
```
Comment (WittierDinosaur) [inline src/sqlfluff/core/linter/linter.py:737]: does this need to be a set of violations? Could we get repeats if we're doing all variants?

File: src/sqlfluff/core/linter/linter.py [hunk 2]
```diff
@@ -714,23 +723,34 @@ def render_string(
...
                 )
             )
+
+        variant_limit = config.get("render_variant_limit")
+        templated_variants: List[TemplatedFile] = []
+        templater_violations: List[SQLTemplaterError] = []
+
         try:
-            templated_file, templater_violations = self.templater.process(
+            for variant, vs in self.templater.process_with_variants(
```
Comment (WittierDinosaur) [inline src/sqlfluff/core/linter/linter.py:732]: can we have a more descriptive variable name than `vs`?

Comment (alanmcruickshank) [reply src/sqlfluff/core/linter/linter.py:732]: yeah - good point

File: src/sqlfluff/core/linter/linter.py (hunk 1 above)
Comment (alanmcruickshank) [reply src/sqlfluff/core/linter/linter.py:737]: yes, we could. I haven't approached the problem of deduping between variants _at all_ yet. I'm going to suggest we don't try and solve that yet - and I'll pick it up as a whole separate challenge at the end, when we can do it for linting too.

I'll add a comment in to make a note on that though, it's a really good point.

File: src/sqlfluff/core/linter/linter.py [hunk 3]
```diff
@@ -303,12 +307,17 @@ def parse_rendered(
...
         tokens: Optional[Sequence[BaseSegment]]
-        if rendered.templated_file is not None:
-            tokens, lvs = cls._lex_templated_file(
-                rendered.templated_file, rendered.config
-            )
+        # TODO: We're limiting ourselves to only the first variant for now.
+        # We'll eventually parse more variants here.
+        if rendered.templated_variants:
+            _root_variant = rendered.templated_variants[0]
+            tokens, lvs = cls._lex_templated_file(_root_variant, rendered.config)
```
Comment (WittierDinosaur) [inline src/sqlfluff/core/linter/linter.py:314]: `lvs` is also pretty ambiguous

Comment (alanmcruickshank) [reply src/sqlfluff/core/linter/linter.py:314]: lol - yeah 🤔 . Will fix that

**Pull request merged**